pkg/provisioner/provisioner.go

```diff
@@ -0,0 +1,111 @@
+/* Copyright 2021 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package provisioner
+
+import (
+	"context"
+	"net/url"
+	"time"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"k8s.io/klog/v2"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+)
+
+const (
+	maxGrpcBackoff  = 30 * time.Second
+	grpcDialTimeout = 30 * time.Second
+)
+
+func NewDefaultCOSIProvisionerClient(ctx context.Context, address string, debug bool) (*COSIProvisionerClient, error) {
+	dialOpts := []grpc.DialOption{
+		grpc.WithInsecure(), // strictly restricting to local Unix domain socket
+		grpc.WithBackoffMaxDelay(maxGrpcBackoff),
+		grpc.WithBlock(), // block until connection succeeds
+		grpc.WithTimeout(grpcDialTimeout),
+	}
+
+	interceptors := []grpc.UnaryClientInterceptor{}
+
+	if debug {
+		interceptors = append(interceptors, apiLogger)
+	}
+	return NewCOSIProvisionerClient(ctx, address, dialOpts, interceptors)
+}
+
+// NewCOSIProvisionerClient creates a new GRPCClient that only supports unix domain sockets
+func NewCOSIProvisionerClient(ctx context.Context, address string, dialOpts []grpc.DialOption, interceptors []grpc.UnaryClientInterceptor) (*COSIProvisionerClient, error) {
+	addr, err := url.Parse(address)
+	if err != nil {
+		return nil, err
+	}
+
+	if addr.Scheme != "unix" {
+		err := errors.New("Address must be a unix domain socket")
+		klog.ErrorS(err, "Unsupported scheme", "expected", "unix", "found", addr.Scheme)
+		return nil, errors.Wrap(err, "Invalid argument")
+	}
+
+	for _, interceptor := range interceptors {
+		dialOpts = append(dialOpts, grpc.WithChainUnaryInterceptor(interceptor))
+	}
+
+	conn, err := grpc.DialContext(ctx, address, dialOpts...)
+	if err != nil {
+		klog.ErrorS(err, "Connection failed", "address", address)
+		return nil, err
+	}
+
+	return &COSIProvisionerClient{
+		address:           address,
+		conn:              conn,
+		identityClient:    cosi.NewIdentityClient(conn),
+		provisionerClient: cosi.NewProvisionerClient(conn),
+	}, nil
+}
+
+func NewDefaultCOSIProvisionerServer(address string,
```

Comment:
Maybe we need to pass a dict of values here instead? Because in the case of COSI we need the endpoint, hmac keys etc and other providers might need other options as well

Reply:
This is the server - this is the core framework. We shouldn't send the driver specific config (like keys) here. This is purely for starting the grpc serer listener.

Reply:
If you want to use driver specific input, pass it in via CLI flags or other mechanisms (like env/metadata server/etc..) and wire it into Provisioner calls directly.